File: frontend/src/services/challengeService.ts
import { api } from "./api";


/* ============================================================
   CHALLENGE EVIDENCE
============================================================ */

export interface ChallengeEvidence {
  id: number;

  evidence_type: string;
  original_filename: string;

  content_type: string | null;
  file_size: number | null;
  file_url: string | null;

  uploaded_by: number;

  created_at: string;
}


/* ============================================================
   CHALLENGE
   Mirrors backend/app/schemas/challenge.py
============================================================ */

export interface Challenge {
  /* ----------------------------------------------------------
     Basic Information
  ---------------------------------------------------------- */

  id: number;

  title: string;
  description: string;

  submitted_by: number;


  /* ----------------------------------------------------------
     Location Resolution
  ---------------------------------------------------------- */

  location_source: string;

  location_verified: boolean;

  location_accuracy_meters: number | null;

  locality: string | null;

  location_resolution_reason: string | null;

  location_resolved_at: string | null;


  /* ----------------------------------------------------------
     Classification
  ---------------------------------------------------------- */

  category: string;

  severity: string;

  urgency: string;


  /* ----------------------------------------------------------
     Impact / Location
  ---------------------------------------------------------- */

  affected_population: number | null;

  estimated_economic_loss: number | null;

  address: string | null;

  district: string | null;

  state: string | null;

  latitude: number | null;

  longitude: number | null;


  /* ----------------------------------------------------------
     Innovation
  ---------------------------------------------------------- */

  innovation_required: boolean;


  /* ----------------------------------------------------------
     General AI Information
  ---------------------------------------------------------- */

  ai_confidence_score: number | null;

  ai_model_version: string | null;


  /* ----------------------------------------------------------
     Category Model
  ---------------------------------------------------------- */

  ai_category_confidence: number | null;

  ai_second_category: string | null;

  ai_second_category_confidence: number | null;

  ai_category_margin: number | null;

  ai_category_decision: string | null;

  ai_requires_human_review: boolean | null;

  ai_category_top_3:
    Array<Record<string, unknown>> | null;


  /* ----------------------------------------------------------
     Innovation-required Model
  ---------------------------------------------------------- */

  ai_innovation_confidence: number | null;

  ai_innovation_decision: string | null;

  ai_innovation_requires_human_review:
    boolean | null;


  /* ----------------------------------------------------------
     Innovation-type Model
  ---------------------------------------------------------- */

  ai_innovation_type: string | null;

  ai_innovation_type_confidence: number | null;

  ai_innovation_type_second: string | null;

  ai_innovation_type_second_confidence:
    number | null;

  ai_innovation_type_margin: number | null;

  ai_innovation_type_decision: string | null;

  ai_innovation_type_requires_human_review:
    boolean | null;

  ai_innovation_type_top_3:
    Array<Record<string, unknown>> | null;


  /* ----------------------------------------------------------
     AI Analysis Timestamp
  ---------------------------------------------------------- */

  ai_analysis_at: string | null;


  /* ----------------------------------------------------------
     Routing
  ---------------------------------------------------------- */

  routing_type: string;

  routing_reason: string | null;


  /* ----------------------------------------------------------
     Current Authority Assignment
  ---------------------------------------------------------- */

  current_authority_id: number | null;

  assigned_at: string | null;

  assigned_by: number | null;


  /* ----------------------------------------------------------
     Challenge Lifecycle
  ---------------------------------------------------------- */

  status: string;


  /* ----------------------------------------------------------
     Duplicate / Master Challenge
  ---------------------------------------------------------- */

  is_master_challenge: boolean;

  master_challenge_id: number | null;

  duplicate_similarity_score: number | null;


  /* ----------------------------------------------------------
     Evidence
  ---------------------------------------------------------- */

  evidence: ChallengeEvidence[];


  /* ----------------------------------------------------------
     Timestamps
  ---------------------------------------------------------- */

  created_at: string;

  updated_at: string;
}


/* ============================================================
   AUTHORITY ROUTING
   Mirrors backend/app/schemas/routing.py
============================================================ */

export interface AuthorityCandidate {
  organization_id: number;

  organization_name: string;

  organization_type: string;

  score: number;

  reason: string;
}


export interface RoutingRecommendation {
  recommended_authority_id: number | null;

  recommended_authority_name:
    string | null;

  score: number;

  reason: string;

  candidates: AuthorityCandidate[];
}


/* ============================================================
   ASSIGNMENT HISTORY
   Mirrors backend/app/schemas/assignment.py
============================================================ */

export interface AssignmentHistoryItem {
  id: number;

  challenge_id: number;

  from_authority_id: number | null;

  to_authority_id: number;

  performed_by: number | null;

  action: string;

  status: string;

  reason: string | null;

  remarks: string | null;

  created_at: string;

  completed_at: string | null;
}


/* ============================================================
   REVIEW HISTORY
   Mirrors backend/app/schemas/review.py
============================================================ */

export interface ReviewHistoryItem {
  id: number;

  challenge_id: number;

  reviewer_id: number;

  decision: string;

  recommended_authority_id: number | null;

  selected_authority_id: number;

  reason: string | null;

  created_at: string;
}


/* ============================================================
   REQUEST TYPES
============================================================ */

export interface ChallengeAssignmentRequest {
  authority_id: number;

  remarks?: string;
}


export interface ChallengeReassignRequest {
  authority_id: number;

  reason: string;

  remarks?: string;
}


export interface ChallengeEscalateRequest {
  authority_id: number;

  reason: string;

  remarks?: string;
}


export interface ReviewOverrideRequest {
  authority_id: number;

  reason: string;
}


/* ============================================================
   CHALLENGE SERVICE
============================================================ */

export const challengeService = {

  /* ----------------------------------------------------------
     GET CHALLENGES
  ---------------------------------------------------------- */

  async getChallenges(
    skip = 0,
    limit = 100,
  ): Promise<Challenge[]> {
    const response =
      await api.get<Challenge[]>(
        "/challenges",
        {
          params: {
            skip,
            limit,
          },
        },
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     GET SINGLE CHALLENGE
  ---------------------------------------------------------- */

  async getChallenge(
    challengeId: number,
  ): Promise<Challenge> {
    const response =
      await api.get<Challenge>(
        `/challenges/${challengeId}`,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     GET MY CHALLENGES
     Useful for citizen-side screens
  ---------------------------------------------------------- */

  async getMyChallenges(): Promise<Challenge[]> {
    const response =
      await api.get<Challenge[]>(
        "/challenges/my",
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     REVIEW QUEUE
  ---------------------------------------------------------- */

  async getReviewQueue(): Promise<Challenge[]> {
    const response =
      await api.get<Challenge[]>(
        "/reviews/queue",
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     ROUTING RECOMMENDATION
  ---------------------------------------------------------- */

  async getRoutingRecommendation(
    challengeId: number,
  ): Promise<RoutingRecommendation> {
    const response =
      await api.get<RoutingRecommendation>(
        `/routing/challenges/${challengeId}/recommendation`,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     ASSIGN CHALLENGE
  ---------------------------------------------------------- */

  async assignChallenge(
    challengeId: number,
    payload: ChallengeAssignmentRequest,
  ): Promise<Challenge> {
    const response =
      await api.post<Challenge>(
        `/challenges/${challengeId}/assign`,
        payload,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     AUTO ROUTE
  ---------------------------------------------------------- */

  async autoRouteChallenge(
    challengeId: number,
  ): Promise<Challenge> {
    const response =
      await api.post<Challenge>(
        `/challenges/${challengeId}/auto-route`,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     REASSIGN CHALLENGE
  ---------------------------------------------------------- */

  async reassignChallenge(
    challengeId: number,
    payload: ChallengeReassignRequest,
  ): Promise<Challenge> {
    const response =
      await api.post<Challenge>(
        `/challenges/${challengeId}/reassign`,
        payload,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     ESCALATE CHALLENGE
  ---------------------------------------------------------- */

  async escalateChallenge(
    challengeId: number,
    payload: ChallengeEscalateRequest,
  ): Promise<Challenge> {
    const response =
      await api.post<Challenge>(
        `/challenges/${challengeId}/escalate`,
        payload,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     ACCEPT AI RECOMMENDATION
  ---------------------------------------------------------- */

  async acceptRecommendation(
    challengeId: number,
  ): Promise<Challenge> {
    const response =
      await api.post<Challenge>(
        `/reviews/${challengeId}/accept`,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     OVERRIDE AI RECOMMENDATION
  ---------------------------------------------------------- */

  async overrideRecommendation(
    challengeId: number,
    payload: ReviewOverrideRequest,
  ): Promise<Challenge> {
    const response =
      await api.post<Challenge>(
        `/reviews/${challengeId}/override`,
        payload,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     ASSIGNMENT HISTORY
  ---------------------------------------------------------- */

  async getAssignmentHistory(
    challengeId: number,
  ): Promise<AssignmentHistoryItem[]> {
    const response =
      await api.get<AssignmentHistoryItem[]>(
        `/challenges/${challengeId}/assignment-history`,
      );

    return response.data;
  },


  /* ----------------------------------------------------------
     REVIEW HISTORY
  ---------------------------------------------------------- */

  async getReviewHistory(
    challengeId: number,
  ): Promise<ReviewHistoryItem[]> {
    const response =
      await api.get<ReviewHistoryItem[]>(
        `/reviews/${challengeId}/history`,
      );

    return response.data;
  },
};

File: backend/app/services/challenge_service.py
from datetime import datetime, timezone
from typing import Any

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models.challenge import (
    Challenge,
    ChallengeLocationSource,
    ChallengeStatus,
)
from app.models.user import User
from app.schemas.challenge import ChallengeCreate, ChallengeUpdate
from app.ai.schemas import TriageRequest
from app.ai.triage_service import triage_challenge
from app.models.challenge_evidence import ChallengeEvidence, EvidenceType
import json


# ============================================================
# Create Challenge
# ============================================================

def create_challenge(
    db: Session,
    challenge_data: ChallengeCreate,
    current_user: User,
) -> Challenge:

    location_resolved_at = None

    if challenge_data.location_source in {
        ChallengeLocationSource.GPS,
        ChallengeLocationSource.GPS_VERIFIED_MANUAL,
        ChallengeLocationSource.CONFLICT,
    }:
        location_resolved_at = datetime.now(timezone.utc)

    challenge = Challenge(
        # ----------------------------------------------------
        # Basic Information
        # ----------------------------------------------------

        title=challenge_data.title,
        description=challenge_data.description,
        submitted_by=current_user.id,

        # ----------------------------------------------------
        # Initial Classification
        # ----------------------------------------------------

        category=challenge_data.category,
        urgency=challenge_data.urgency,

        # ----------------------------------------------------
        # Impact
        # ----------------------------------------------------

        affected_population=challenge_data.affected_population,
        estimated_economic_loss=(
            challenge_data.estimated_economic_loss
        ),

        # ----------------------------------------------------
        # Location
        # ----------------------------------------------------

        address=challenge_data.address,
        district=challenge_data.district,
        state=challenge_data.state,
        locality=challenge_data.locality,

        latitude=challenge_data.latitude,
        longitude=challenge_data.longitude,

        # ----------------------------------------------------
        # Location Resolution Metadata
        # ----------------------------------------------------

        location_source=challenge_data.location_source,
        location_verified=challenge_data.location_verified,
        location_accuracy_meters=(
            challenge_data.location_accuracy_meters
        ),
        location_resolution_reason=(
            challenge_data.location_resolution_reason
        ),
        location_resolved_at=location_resolved_at,
    )

    db.add(challenge)
    db.commit()
    db.refresh(challenge)

    if challenge_data.media_ids:
        for media_id_str in challenge_data.media_ids:
            try:
                metadata = json.loads(media_id_str)
                evidence = ChallengeEvidence(
                    challenge_id=challenge.id,
                    evidence_type=EvidenceType(metadata["evidence_type"]),
                    original_filename=metadata["original_filename"],
                    stored_filename=metadata["stored_filename"],
                    content_type=metadata["content_type"],
                    file_size=metadata["file_size"],
                    file_url=metadata["file_url"],
                    uploaded_by=current_user.id
                )
                db.add(evidence)
            except Exception as e:
                # Log error or continue
                print(f"Failed to process media_id {media_id_str}: {e}")
        db.commit()

    return challenge


# ============================================================
# Get Challenge
# ============================================================

def get_challenge_by_id(
    db: Session,
    challenge_id: int,
) -> Challenge | None:
    statement = select(Challenge).where(
        Challenge.id == challenge_id
    )

    challenge = db.scalar(statement)
    
    if challenge:
        _ensure_ai_analysis(db, [challenge])
        
    return challenge


# ============================================================
# Dynamic AI Analysis Helper
# ============================================================

def _ensure_ai_analysis(db: Session, challenges: list[Challenge]):
    for challenge in challenges:
        if challenge.ai_analysis_at is None:
            # Construct a request for the local AI models
            request = TriageRequest(
                title=challenge.title,
                description=challenge.description,
                category=challenge.category,
                address=challenge.address,
                district=challenge.district,
                state=challenge.state,
                affected_population=challenge.affected_population,
                estimated_economic_loss=challenge.estimated_economic_loss,
            )
            # Run triage locally
            triage_result = triage_challenge(request)
            # Persist it
            persist_ai_triage_result(db, challenge, triage_result)


# ============================================================
# Get Challenges
# ============================================================

def get_challenges(
    db: Session,
    skip: int = 0,
    limit: int = 20,
) -> list[Challenge]:

    statement = (
        select(Challenge)
        .order_by(Challenge.created_at.desc())
        .offset(skip)
        .limit(limit)
    )

    challenges = list(db.scalars(statement).all())
    _ensure_ai_analysis(db, challenges)
    return challenges


# ============================================================
# Get User Challenges
# ============================================================

def get_user_challenges(
    db: Session,
    user_id: int,
    skip: int = 0,
    limit: int = 20,
) -> list[Challenge]:

    statement = (
        select(Challenge)
        .where(
            Challenge.submitted_by == user_id
        )
        .order_by(
            Challenge.created_at.desc()
        )
        .offset(skip)
        .limit(limit)
    )

    challenges = list(db.scalars(statement).all())
    _ensure_ai_analysis(db, challenges)
    return challenges


# ============================================================
# Update Challenge
# ============================================================

def update_challenge(
    db: Session,
    challenge: Challenge,
    challenge_data: ChallengeUpdate,
) -> Challenge:

    update_data = challenge_data.model_dump(
        exclude_unset=True
    )

    for field, value in update_data.items():
        setattr(
            challenge,
            field,
            value,
        )

    db.commit()
    db.refresh(challenge)

    return challenge


# ============================================================
# JSON Safety Helper
# ============================================================

def _json_safe(value: Any):
    """
    Convert Pydantic models and nested structures into
    JSON-serializable Python primitives.
    """

    if value is None:
        return None

    if hasattr(value, "model_dump"):
        return value.model_dump()

    if hasattr(value, "dict"):
        return value.dict()

    if isinstance(value, list):
        return [
            _json_safe(item)
            for item in value
        ]

    if isinstance(value, tuple):
        return [
            _json_safe(item)
            for item in value
        ]

    if isinstance(value, dict):
        return {
            key: _json_safe(val)
            for key, val in value.items()
        }

    return value


# ============================================================
# Persist AI Triage Result
# ============================================================

def persist_ai_triage_result(
    db: Session,
    challenge: Challenge,
    triage_result,
) -> Challenge:
    """
    Persist the complete AI triage result.

    Category, severity, urgency, innovation, routing,
    model evidence, confidence values, and explanations
    are persisted for auditability and frontend display.
    """

    # ========================================================
    # CANONICAL TRIAGE VALUES
    # ========================================================

    challenge.category = (
        triage_result.category
    )

    challenge.severity = (
        triage_result.severity
    )

    challenge.urgency = (
        triage_result.urgency
    )

    challenge.innovation_required = (
        triage_result.innovation_required
    )

    # ========================================================
    # MODEL VERSION / GENERAL AI CONFIDENCE
    # ========================================================

    challenge.ai_confidence_score = (
        triage_result.category_confidence
    )

    challenge.ai_model_version = (
        triage_result.model_version
    )

    # ========================================================
    # CATEGORY AI EVIDENCE
    # ========================================================

    challenge.ai_category_confidence = (
        triage_result.category_confidence
    )

    challenge.ai_second_category = (
        triage_result.second_category
    )

    challenge.ai_second_category_confidence = (
        triage_result.second_category_confidence
    )

    challenge.ai_category_margin = (
        triage_result.category_margin
    )

    challenge.ai_category_decision = (
        triage_result.category_decision
    )

    challenge.ai_requires_human_review = (
        triage_result.requires_human_review
    )

    challenge.ai_category_top_3 = _json_safe(
        triage_result.category_top_3
    )

    # ========================================================
    # INNOVATION REQUIRED AI EVIDENCE
    # ========================================================

    challenge.ai_innovation_confidence = (
        triage_result.innovation_confidence
    )

    challenge.ai_innovation_decision = (
        triage_result.innovation_decision
    )

    challenge.ai_innovation_requires_human_review = (
        triage_result.innovation_requires_human_review
    )

    # ========================================================
    # INNOVATION TYPE AI EVIDENCE
    # ========================================================

    challenge.ai_innovation_type = (
        triage_result.innovation_type
    )

    challenge.ai_innovation_type_confidence = (
        triage_result.innovation_type_confidence
    )

    challenge.ai_innovation_type_second = (
        triage_result.innovation_type_second
    )

    challenge.ai_innovation_type_second_confidence = (
        triage_result.innovation_type_second_confidence
    )

    challenge.ai_innovation_type_margin = (
        triage_result.innovation_type_margin
    )

    challenge.ai_innovation_type_decision = (
        triage_result.innovation_type_decision
    )

    challenge.ai_innovation_type_requires_human_review = (
        triage_result.innovation_type_requires_human_review
    )

    challenge.ai_innovation_type_top_3 = _json_safe(
        triage_result.innovation_type_top_3
    )

    # ========================================================
    # ANALYSIS TIMESTAMP
    # ========================================================

    challenge.ai_analysis_at = (
        datetime.now(timezone.utc)
    )

    # ========================================================
    # ROUTING
    # ========================================================

    challenge.routing_type = (
        triage_result.routing_type
    )

    challenge.routing_reason = (
        triage_result.routing_reason
    )

    # ========================================================
    # FINAL CHALLENGE STATUS
    # ========================================================

    if triage_result.requires_human_review:
        challenge.status = (
            ChallengeStatus.UNDER_REVIEW
        )
    else:
        challenge.status = (
            ChallengeStatus.ROUTED
        )

    db.add(challenge)
    db.commit()
    db.refresh(challenge)

    return challenge